File: root/src/kafka_producer.py
from pyspark.sql import SparkSession
from pyspark.sql.functions import to_json, struct, from_xml
from pyspark.streaming import StreamingContext
from pyspark.streaming.kafka import KafkaUtils

spark = SparkSession.builder.appName("KafkaProducer").getOrCreate()

# Below is the schema definition for xml data 
xml_schema = StructType([
    StructField("name", StringType(), True),
    StructField("date", StringType(), True),
    StructField("sever-details", StructType([
        StructField("server_id", IntegerType(), True),
        StructField("location", StringType(), True),
        StructField("temp", FloatType(), True),
    ]), True),
])
 
# Reading XML data with schema
df = (
    spark.readStream
    .format("xml")
    .option("rowTag", "event")
    .schema(xml_schema)
    .load("/Desktop/stream_file.xml")  
)

# Select necessary columns and convert DataFrame to JSON
df_json = df.select(
    to_json(struct("name", "date", "server-details")).alias("value")
)

# Configure Kafka producer parameters
kafka_params = {
    "bootstrap.servers": "localhost:9092",
    "key.serializer": "org.apache.kafka.common.serialization.StringSerializer",
    "value.serializer": "org.apache.kafka.common.serialization.StringSerializer"
}

# kafka topic 
kafka_topic = "events"

# Publishing data to Kafka topic in a streaming manner
query = (
    df_json.writeStream
    .format("kafka")
    .outputMode("append")
    .option("kafka.bootstrap.servers", kafka_params["bootstrap.servers"])
    .option("topic", "events")
    .start()
)

query.awaitTermination()

File: root/scala/kafka_producer.scala
import org.apache.spark.sql.{SparkSession, Row}
import org.apache.spark.sql.functions._
import org.apache.spark.sql.types._
import org.apache.spark.streaming.{StreamingContext, _}
import org.apache.spark.streaming.kafka._

val spark = SparkSession.builder.appName("KafkaProducer").getOrCreate()

//  XML schema definition
val xmlSchema = StructType(Seq(
  StructField("name", StringType, nullable = true),
  StructField("date", StringType, nullable = true),
  StructField("sever-details", StructType(Seq(
    StructField("server_id", IntegerType, nullable = true),
    StructField("location", StringType, nullable = true),
    StructField("temp", FloatType, nullable = true)
  )), nullable = true)
))

// Read XML data with schema
val df = spark
  .readStream
  .format("xml")
  .option("rowTag", "event")
  .schema(xmlSchema)
  .load("/Desktop/stream_file.xml")

//convert to JSON
val dfJson = df.select(
  to_json(struct("name", "date", "server-details")).alias("value")
)

val kafkaParams = Map(
  "bootstrap.servers" -> "localhost:9092",
  "key.serializer" -> "org.apache.kafka.common.serialization.StringSerializer",
  "value.serializer" -> "org.apache.kafka.common.serialization.StringSerializer"
)

val kafkaTopic = "events"

// Write data to Kafka topic in streaming
val query = dfJson
  .writeStream
  .format("kafka")
  .outputMode("append")
  .option("kafka.bootstrap.servers", kafkaParams("bootstrap.servers"))
  .option("topic", kafkaTopic)
  .start()

query.awaitTermination()
